widgetssdk/src/main/java/com/glia/widgets/view/textview/ChatInputHintFix.kt

```diff
@@ -0,0 +1,87 @@
+import android.text.Editable
+import android.text.TextUtils
+import android.text.TextWatcher
+import android.util.AttributeSet
+
+/**
+ * This class is here to solve next bug:
+ * See video in this comment to understand initial bug: https://glia.atlassian.net/browse/MOB-3860?focusedCommentId=459636
+ *
+ * The issue was caused by default EditText behavior. The hint height is effecting the EditText height even when the hint is not displayed.
+ * This means that of for some reason your EditText view has property maxLines = 4 and your hint is so long that it takes 4 lines and
+ * then when user will wright just 1 letter into your EditText it would still as much space as if it had 4 lines of code.
+ *
+ * Another related problem is that for some reason at least in ConstraintLayout multi-line hint messes up layout rules. See next comment
+ * in the same ticket: https://glia.atlassian.net/browse/MOB-3860?focusedCommentId=459642.
+ * To prevent this from happening this implementation forces hint to always be one-line
+ */
+internal class ChatInputHintFix : androidx.appcompat.widget.AppCompatEditText {
+
+    constructor(context: Context) : super(context) {
+        init(context, null)
+    }
+
+    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
+        init(context, attrs)
+    }
+
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
+        init(context, attrs)
+    }
+
+    private var hintOverride: CharSequence? = null
+    var maxLinesOverride: Int = 0
+    var ellipsizeOverride: TextUtils.TruncateAt? = null
+
+    private fun init(context: Context, attrs: AttributeSet?) {
+        hintOverride = hint           // Retrieves existing value if it was set in XML
+        maxLinesOverride = maxLines   // Retrieves existing value if it was set in XML
+        ellipsizeOverride = ellipsize // Retrieves existing value if it was set in XML
+
+        if (shouldShowHint()) {
+            showHintAndMakeItOneLine()
+        }
+        addTextChangedListener(object : TextWatcher {
+            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
+                /* no-op */
+            }
+
+            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
+                /* no-op */
+            }
+
+            override fun afterTextChanged(s: Editable?) {
+                if (shouldShowHint()) {
+                    showHintAndMakeItOneLine()
+                } else if (!s.isNullOrEmpty() && !hint.isNullOrEmpty()) {
+                    removeHintAndRestoreDefaultSettings()
+                }
+            }
+
+        })
+    }
+
+    fun setHintOverride(s: CharSequence?) {
+        hintOverride = s
+        if (shouldShowHint()) {
+            showHintAndMakeItOneLine()
+        }
+    }
+
+    private fun shouldShowHint(): Boolean = text.isNullOrEmpty() && hint != hintOverride
```

Comment:
Could you please explain the `&& hint != hintOverride` part? I'm not sure I understand it.

Reply:
I am not sure what exactly is confusing so I will try to note down everything that comes to mind:
- `hint` here is the native EditText#getHint() method
- when the user has written anything into this EditText then the hint is set to be empty (e.g.: `""`) to prevent a bug with height but later we need to add it back when there is no user text in this EditText
- Technically I could have used just the part `text.isNullOrEmpty()` but I wanted to make sure that it is not triggered unnecessarily if the hint is already displayed
- also it is possible that the hint has been changed at runtime using the setHintOverride (e.g. custom locales finished loading and need hint update) while there was an old hint displayed needing refresh

Reply:
That's a great explanation! I think it makes sense even to put it as comments to appropriate places of `ChatInputHintFix.kt`. 

> also it is possible that the hint has been changed at runtime using the setHintOverride (e.g. custom locales finished loading and need hint update) while there was an old hint displayed needing refresh

I was missing this point. I couldn't figure out when hint `might` not match the `hintOverride` and why it influences the `shouldShowHint()` result.

Reply:
Fixed in 7dc2b11